core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/devui/ChatJsonRPCService.java

```diff
@@ -225,12 +231,27 @@ public ChatResultPojo chat(String message, boolean ragEnabled) {
                 memory.add(new UserMessage(message));
             }
 
+            boolean hasToolProvider = toolProvider != null;
+            if (hasToolProvider) {
+                ToolProviderRequest toolRequest = new ToolProviderRequest(memory, userMessage);
+                ToolProviderResult toolsResult = toolProvider.provideTools(toolRequest);
+                for (ToolSpecification specification : toolsResult.tools().keySet()) {
+                    toolSpecifications.add(specification);
+                    toolExecutors.put(specification.name(), toolsResult.tools().get(specification));
+                }
+            }
+
             Response<AiMessage> modelResponse;
             if (toolSpecifications.isEmpty()) {
                 modelResponse = model.generate(memory.messages());
                 memory.add(modelResponse.content());
             } else {
-                modelResponse = executeWithTools(memory);
+                executeWithTools(memory);
+            }
+            // Remove toolProviderSupplier tools again
+            if (hasToolProvider) {
+                toolSpecifications.clear();
```

Comment:
IIUC, if there's a class with static tools (with annotations) AND a dynamic tool provider, then when the `ChatJsonRPCService` is initialized, then information from the "static" tools is added and made available to the model, but then after handling the first request, you clear the `toolSpecifications` and `toolExecutors`, and on the second request you only fill it again with the dynamic tools, but not the static tools (these are lost for subsequent chatting). I'm not completely sure if we want to allow mixing both of these "tool sources" in the Dev UI (it's not allowed on AI services, right?), but maybe it wouldn't hurt to pick up and combine both in this case?! 
So, maybe these `toolSpecifications` and `toolExecutors` fields should only contain the static tools, and the dynamic ones should be retrieved and handled via local variables in the chat methods. WDYT?

Reply:
Another thing is that you don't support streaming chats right now, so after splitting `toolSpecifications` and `toolExecutors` into separate dynamic and static ones, we could also retrieve and apply the dynamic ones in the `streamingChat` method

Reply:
Thanks for the feedback.
I will work on streamingChat implementation, today then. 
About mixing static tools and dynamic (toolProvider) tools. In my opinion, it is more confusing to mix them in the DevUI `ChatJsonRPCService`. The `AiService` doesn't allow that, and it would be inconsistent.
I will try out if the DevUI starts with toolProvider and static tools. If the DevUI starts, I will need to fix that.

Reply:
Yeah it should either use both sources, or always only one, but if it picks up both on the first request and only dynamic tools on subsequent requests, that's wrong.

Reply:
You are right. I will fix this :)

Reply:
The question is, should I ignore static tools when dynamic tools are found or throw an error?
Also, in the "Tools" tab, I only show a list of `ToolProvider`, if any are discovered.

Reply:
Hmm, it depends on how we think users will use the application. There are two options

- A: There will be a set of static tools, and a dynamic provider that provides some *extra* methods that do not correspond to any of the static tools. In this case, it would make sense to let the application use both (also an AI service should be able to use both, and so should the Dev UI page)
- B: There will be a set of static tools, and the dynamic provider's main use will be to "filter" which tools to make available at a given moment (it will return references to some of the existing static tool methods). In this case, generally only the dynamic provider should be picked up (in the Dev UI), because by picking up both, you would get duplicate tools

I assume your main target when implementing this was to support case B, right?

Reply:
Or maybe C, where the app should have either only static, or only dynamic tools. If that's the case, we should prefer dynamic tools only, and ignore the static ones.

Reply:
Option C sounds right. In Langchain4J, we specified, that [either static tools or the tool provider can be configured, but not both](https://github.com/langchain4j/langchain4j/pull/1644#discussion_r1732961279) for an AiService. My last [commit 1fb3761](https://github.com/quarkiverse/quarkus-langchain4j/pull/989/commits/1fb3761fe83544df025e9f2ef07ef598fe338ce2) follows this specification :)
If a toolProvider is found, we ignore static tools and only use the provider (dynamic tools).

Reply:
Ok then! Let's just make that clear in the quarkus-langchain4j docs too, because as you see, I understood it differently at first :)